crates/sui-data-ingestion-core/src/reducer.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{Reducer, Worker, MAX_CHECKPOINTS_IN_PROGRESS};
+use anyhow::Result;
+use futures::StreamExt;
+use std::collections::HashMap;
+use sui_types::messages_checkpoint::CheckpointSequenceNumber;
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::ReceiverStream;
+
+pub(crate) async fn reduce<W: Worker>(
+    task_name: String,
+    mut current_checkpoint_number: CheckpointSequenceNumber,
+    progress_receiver: mpsc::Receiver<(CheckpointSequenceNumber, W::Result)>,
+    executor_progress_sender: mpsc::Sender<(String, CheckpointSequenceNumber)>,
+    reducer: Option<Box<dyn Reducer<W::Result>>>,
+) -> Result<()> {
+    let mut stream =
+        ReceiverStream::new(progress_receiver).ready_chunks(MAX_CHECKPOINTS_IN_PROGRESS);
+    let mut unprocessed = HashMap::new();
+    let mut batch = vec![];
+
+    while let Some(progress_update) = stream.next().await {
+        for (checkpoint_number, message) in progress_update {
+            unprocessed.insert(checkpoint_number, message);
+        }
+        while let Some(message) = unprocessed.remove(&current_checkpoint_number) {
+            if reducer.is_some() {
+                batch.push(message);
+            }
+            current_checkpoint_number += 1;
+        }
+        let new_watermark = match reducer {
+            Some(ref reducer) => {
+                if !batch.is_empty() {
+                    batch = reducer.partial_commit(std::mem::take(&mut batch)).await?;
```

Comment:
There are a few efficiency problems in this implementation.
First of all, it started the committing step only after receiving one item from the stream. This can be very inefficient, as it may end up committing one item at a time. Unless you require all implementers to think about adding a lower bound on the batch size before committing, which is not very natural for most pipelines.
Another issue is that the remaining batch is not committed again. In most pipelines, they would want to keep committing until the batch is empty, since all they want is to chunk the batch to avoid overwhelming the DB. The archival pipeline is actually a rare exception where it doesn't want to commit the batch if it's not big enough.
So I think an ideal efficient implementation should look something like this:
- An async thread loops receiving items and putting them in a DashMap.
- Main thread wakes up every fixed interval, optionally defined by the pipeline, and start removing items from the DashMap in sequence number order one by one. After each removal, it checks a function to decide whether it's ready for commit, if yes, call the commit function; otherwise continue.

Reply:
>  it started the committing step only after receiving one item from the stream

how is it different from existing [stream](https://github.com/MystenLabs/sui/blob/8761bf8b70de387bf9ec35025493688482793c60/crates/sui-indexer/src/handlers/committer.rs#L46) for a primary workflow? It commits what's available. If you want to accumulate more you can return full vector back in partial_commit. We can also add some lower bound method in trait if that's helpful

> Another issue is that the remaining batch is not committed again

not sure I get this part. You have an option to commit any part of accumulated batch. By default you commit everything and there's no remaining part. You can also split into chunks and make subcommits inside of that call. 

> Main thread wakes up every fixed interval

all progress messages are going through this actor(even with no reducer defined); we don't want to introduce any kind of extra latencies here in updating progress

Reply:
The speculation was that, in most cases, users would prefer to batch as much as possible and commit it all at once, requiring implementing just one method for that. And more nuanced scenarios could be addressed by the partial_commit method. The speculation might be completely off btw:)

Overall I think that both the `commit + partial_commit` approach and your interface suggestion are better than the original version. I don’t have a strong preference, so I’m ok to change it

Reply:
Looking at the current main indexer pipeline, all commits are batched though. My understanding is that there is some limit in PG (e.g. @amnn just hit this in a backfill: https://github.com/MystenLabs/sui/pull/19699) that most likely some basic batching needs to be in place for all pipelines.

Reply:
yes, but I don't see a lower limit for batch size in terms of the number of checkpoints, which makes sense. When you're caught up and tailing the live checkpoint stream, you want to ingest as fast as possible. During a backfill though, the size of batches will start growing by itself because more checkpoints will accumulate before the reducer processes the previous batch

Reply:
Not sure if I understand, but I thought the data that we pass to `commit` is already the processed data and the length of that data should just be the number of rows that we would commit to DB.
So that definitely need to be capped? Or did I confuse about the semantics of the `commit` function.

Reply:
oh, sorry! We were discussing different things. I was referring to a lower limit, the upper limit makes a lot of sense.
Updated the PR to `should_close_batch` interface